### ISSUE TITLE
asterisk key only types asterisk 

### ISSUE THREAD

**Issue by paulopaniago** [nimble0/dotterel#34]
When I type the asterisks keys, it only imprint the asterisk symbol *, instead of reverting the last key stroke. Is that how Dotterel should work?
I am using Samsung Tab Active 3 with Android 13, and this behavior happens in all windows I have tried (wps, samsung notes and chrome)

**Comment by nimble0** [comment]
It sounds like you're using your own dictionary. Dotterel requires a specific entry for undo. Try adding:

`"*": "{RETRO:UNDO}"`

Alternatively you could add the [built-in commands dictionary](https://github.com/nimble0/dotterel/blob/master/app/src/main/assets/dictionaries/commands.json) to your dictionary list.

**Comment by paulopaniago** [comment]
Hey, that was exactly the problem, thank you so much!
This program is priority in my life now. Since I recently had some back issues, I am having problems carrying my laptop, so I required a lighter tablet instead, and Dotterel might just be a life saver. I intend to use this tablet with the keyboard basically like a writing machine, it doesn't need many capabilities, it just need to write with steno on a text editor. 
Because of that, I made a specific dictionary for Dotterel, since I type in Portuguese and it has a different layout then the English version. I'll leave in here, in case some day you want to add this feature.
(Portugues steno layout)
SKFL*RBGTD
 STPR*WPHSZ
     AOEU
Basically, I made a dictionary in which I type correctly, and the translation goes correctly also, but the steno codes are "wrong" since it is for the English layout. I had already made a dictionary like that ounce, many years ago when Plover only had the English layout, so I sort of new what to do.
Also, I'm glad to tell you that Dotterel works over Bluetooth, at least it does on the Corsair K70 mini.

**Comment by nimble0** [comment]
I hope it's able to meet your needs.

It is possible to use different key layouts. You'll have to directly edit the system's JSON file. For reference, [this is the English steno system JSON](https://github.com/nimble0/dotterel/blob/master/app/src/main/assets/systems/ireland.english.json).

`keyLayout` specifies the available steno keys.

`machineConfig` specifies how the steno keys map to machine keys. You can just swap the English steno keys with whichever key it's been replaced by.

You'll probably want to change `prefixStrokes` and `suffixStrokes` too. In English these are the strokes that correspond to things like "ing", "ed", etc.

For a different language you'll want to remove the English orthography files.

Here's a full example:
```
{
	"base": "asset:/systems/ireland.english.json",
	"keyLayout": {
		"keysString": "#1S2KT3FP4LR-5A0O*EU-6RW7BP8GH9TSDZ",
		"mappedKeys": {
			"1-": ["#-", "S-"],
			"2-": ["#-", "K-"],
			"3-": ["#-", "F-"],
			"4-": ["#-", "L-"],
			"5": ["#-", "A-"],
			"0": ["#-", "O-"],
			"-6": ["#-", "-"],
			"-7": ["#-", "-P"],
			"-8": ["#-", "-L"],
			"-9": ["#-", "-T"]
		}
	},
	"prefixStrokes": [],
	"suffixStrokes": [],

	"orthographies": [],

	"dictionaries": [
		{"path": "code_dictionary:/Numbers", "enabled": true}
	],

	"machineConfig": {
		"NKRO": {
			"layout": {
				"#-": ["1", "2", "3", "4", "5", "6", "7", "8", "9", "0", "minus", "equals"],
				"S-": ["q", "a"],
				"K-": ["w"],
				"T-": ["s"],
				"F-": ["e"],
				"P-": ["d"],
				"L-": ["r"],
				"R-": ["f"],
				"A": ["c"],
				"O": ["v"],
				"*": ["t", "g", "y", "h"],
				"E": ["n"],
				"U": ["m"],
				"-R": ["u"],
				"-W": ["j"],
				"-B": ["i"],
				"-P": ["k"],
				"-G": ["o"],
				"-H": ["l"],
				"-T": ["p"],
				"-S": ["semicolon"],
				"-D": ["left_bracket"],
				"-Z": ["apostrophe"]
			}
		},
		"On Screen": {
			"styles": {
				"Touch": {
					"viewLayout": "Ireland",
					"layout": {
						"#-": ["#-"],
						"S-": ["S-"],
						"K-": ["T-"],
						"T-": ["K-"],
						"F-": ["P-"],
						"P-": ["W-"],
						"L-": ["H-"],
						"R-": ["R-"],
						"A": ["A-"],
						"O": ["O-"],
						"*": ["*"],
						"E": ["-E"],
						"U": ["-U"],
						"-R": ["-F"],
						"-W": ["-R"],
						"-B": ["-P"],
						"-P": ["-B"],
						"-G": ["-L"],
						"-H": ["-G"],
						"-T": ["-T"],
						"-S": ["-S"],
						"-D": ["-D"],
						"-Z": ["-Z"]
					}
				},
				"Swipe": {
					"viewLayout": "Ireland",
					"layout": {
						"#-": ["#-"],
						"S-": ["S-"],
						"K-": ["T-"],
						"T-": ["K-"],
						"F-": ["P-"],
						"P-": ["W-"],
						"L-": ["H-"],
						"R-": ["R-"],
						"A": ["A-"],
						"O": ["O-"],
						"*": ["*"],
						"E": ["-E"],
						"U": ["-U"],
						"-R": ["-F"],
						"-W": ["-R"],
						"-B": ["-P"],
						"-P": ["-B"],
						"-G": ["-L"],
						"-H": ["-G"],
						"-T": ["-T"],
						"-S": ["-S"],
						"-D": ["-D"],
						"-Z": ["-Z"]
					}
				}
			}
		}
	}
}
```

**Comment by paulopaniago** [comment]
Hey man, that is valuable information. I'm still trying to contact someone to help me with that, since I'm very obtuse to do anything related to programming.  I guess I just need to learn how to open the application, modify the layout and steno order and then somehow turn it into an application again to install it on the tablet.

**Comment by nimble0** [comment]
Oh no, you don't need to edit the application. You just navigate in the Dotterel settings application to `Systems > My System > JSON`.

**Comment by paulopaniago** [comment]
Oh it is working now! Thank you so much Nimble! I have only realized that you had already made that adaptation in you "full example", after I tried it out making some mistakes of my own. That is really cool! Even the virtual keyboard is working in Portuguese! 
And may I ask you another question?
Is there a way to add a suffix rule in which it erases the last vowel an replaces it with the suffix? 
Here is the example in Plover, and it is the only orthographic rule for Plover for Portuguese so far, this is the only difference now that still exists between Plover and Dotterel, when using it for Portuguese:
passo + {^ando} = passando
massa + {^agem} = massagem
fera + {^ido} = ferido
tal + {^o} = talo 
(notice that this rule doesn't erase consonants, only vowels)

Here below is an example I've copied from the Discord discussion where Ted Morin and Bernoit Pierre were adding this feature:

 plover_portuguese.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git i/plover_portuguese.py w/plover_portuguese.py
index 4faf83c..af455a9 100755
--- i/plover_portuguese.py
+++ w/plover_portuguese.py
@@ -34,7 +34,7 @@ ORTHOGRAPHY_RULES = [
     # Collapse vowels in suffixes
     # como + endo = comendo
     # cai + iria = cairia
-    (r'^(.+)[aeouiáéíóúãõâêôàü] \^ ([aeouiáéíóúãõâêôàü].+)$', r'\1\2'),
+    (r'^(.+)[aeouiáéíóúãõâêôàü] \^ ([aeouiáéíóúãõâêôàü]\w*)$', r'\1\2'),
 ]
 
 ORTHOGRAPHY_RULES_ALIASES = {}

**Comment by nimble0** [comment]
Add this orthography file:
[portuguese.regex.json](https://github.com/user-attachments/files/18146669/portuguese.regex.json)

There isn't an option to add orthography file so add it as a dictionary file then go into the JSON and move it from the dictionaries section to the orthographies section.

So change from something like this:
```
"orthographies": [
],
"dictionaries": [
	{
		"path": "code_dictionary:/Numbers",
		"enabled": true
	},
	{
		"path": "content://portuguese.regex.json",
		"enabled": true
	}
]
```
to this:
```
"orthographies": [
	{
		"path": "content://portuguese.regex.json",
		"enabled": true
	}
],
"dictionaries": [
	{
		"path": "code_dictionary:/Numbers",
		"enabled": true
	}
]
```
Make sure there's a no trailing comma after the last dictionary entry.

**Comment by paulopaniago** [comment]
Hello my friend, unfortunately I couldn't make it work. I tried it for a few times and was going to try it today again, but ended up not doing anything.
For some reason the portuguese.regex.json file was not readable by my tablet, so I erased the [ ] and added this bellow as another dictionary (.json) in Dotterel:

{
"l": "[aeouiáéíóúãõâêôàü]", "r": "([aeouiáéíóúãõâêôàü])", "s": "$1"
}


And here is what I did trying to follow your orientation:

{
	"base": "asset:/systems/ireland.english.json",
	"keyLayout": {
		"keysString": "#1S2KT3FP4LR-5A0O*EU-6RW7BP8GH9TSDZ",
		"mappedKeys": {
			"1-": ["#-", "S-"],
			"2-": ["#-", "K-"],
			"3-": ["#-", "F-"],
			"4-": ["#-", "L-"],
			"5": ["#-", "A-"],
			"0": ["#-", "O-"],
			"-6": ["#-", "-"],
			"-7": ["#-", "-P"],
			"-8": ["#-", "-L"],
			"-9": ["#-", "-T"]
		}
	},
	"prefixStrokes": [],
	"suffixStrokes": [],

	"orthographies": [
	{
		"path": "content://portuguese.regex.json",
		"enabled": true
	}
],
"dictionaries": [
	{
		"path": "code_dictionary:/Numbers",
		"enabled": true
	}
],

	"machineConfig": {
		"NKRO": {
			"layout": {
				"#-": ["1", "2", "3", "4", "5", "6", "7", "8", "9", "0", "minus", "equals"],
				"S-": ["q", "a"],
				"K-": ["w"],
				"T-": ["s"],
				"F-": ["e"],
				"P-": ["d"],
				"L-": ["r"],
				"R-": ["f"],
				"A": ["c"],
				"O": ["v"],
				"*": ["t", "g", "y", "h"],
				"E": ["n"],
				"U": ["m"],
				"-R": ["u"],
				"-W": ["j"],
				"-B": ["i"],
				"-P": ["k"],
				"-G": ["o"],
				"-H": ["l"],
				"-T": ["p"],
				"-S": ["semicolon"],
				"-D": ["left_bracket"],
				"-Z": ["apostrophe"]
			}
		},
		"On Screen": {
			"styles": {
				"Touch": {
					"viewLayout": "Ireland",
					"layout": {
						"#-": ["#-"],
						"S-": ["S-"],
						"K-": ["T-"],
						"T-": ["K-"],
						"F-": ["P-"],
						"P-": ["W-"],
						"L-": ["H-"],
						"R-": ["R-"],
						"A": ["A-"],
						"O": ["O-"],
						"*": ["*"],
						"E": ["-E"],
						"U": ["-U"],
						"-R": ["-F"],
						"-W": ["-R"],
						"-B": ["-P"],
						"-P": ["-B"],
						"-G": ["-L"],
						"-H": ["-G"],
						"-T": ["-T"],
						"-S": ["-S"],
						"-D": ["-D"],
						"-Z": ["-Z"]
					}
				},
				"Swipe": {
					"viewLayout": "Ireland",
					"layout": {
						"#-": ["#-"],
						"S-": ["S-"],
						"K-": ["T-"],
						"T-": ["K-"],
						"F-": ["P-"],
						"P-": ["W-"],
						"L-": ["H-"],
						"R-": ["R-"],
						"A": ["A-"],
						"O": ["O-"],
						"*": ["*"],
						"E": ["-E"],
						"U": ["-U"],
						"-R": ["-F"],
						"-W": ["-R"],
						"-B": ["-P"],
						"-P": ["-B"],
						"-G": ["-L"],
						"-H": ["-G"],
						"-T": ["-T"],
						"-S": ["-S"],
						"-D": ["-D"],
						"-Z": ["-Z"]
					}
				}
			}
		}
	}
}

**Comment by nimble0** [comment]
Did you replace the path `content://portuguese.regex.json` with whatever path it gave when you added it as a dictionary?

**Comment by paulopaniago** [comment]
I'm sorry, I didn't understand the question. All I did was to change the
example you wrote, and import the portuguese.regex.json to the dictionaries
section of Dotterel. Since it wasn´t readable, giving that red color
warning, I changed the [ ] for { }. But it still did not modify its
behavior.
For now I am using the command {#BackSpace}{^^} for it to erase the last
letter and replace it with the suffix. It is not as good as an orthographic
rule, since it doesn't discriminate between consonants or vowels. I'm
sending my dictionaries with suffixes, if you would like to check it. The
one I use in Dotterel and the equivalent in Plover.


Em dom., 22 de dez. de 2024 às 05:18, nimble0 ***@***.***>
escreveu:

**Comment by nimble0** [comment]
Ah, no, `portuguese.regex.json` won't be readable as a dictionary file. Adding as a dictionary file is only for getting the path and getting permission to read the file.

**Comment by paulopaniago** [comment]
Hello Nimble, I tried a few more times to make the orthographic rules work after I came from vacations, with no success. I guess I am probably doing wrong, something that should be very simple for others. But I given up on this for now. Dotterel is actually working fine with those commands, instead of the orthographic rules. I used it daily in the last two weeks, and I can't express how much I am grateful for your work my friend!
